pkg/services/provisioning/provisioning.go

```diff
@@ -17,31 +18,83 @@ func init() {
 }
 
 type ProvisioningService struct {
-	Cfg *setting.Cfg `inject:""`
+	Cfg                 *setting.Cfg `inject:""`
+	log                 log.Logger
+	dashProvisionerChan chan *dashboards.DashboardProvisioner
+	pauseChan           chan interface{}
+	running             bool
+
+	runCtxCancel context.CancelFunc
 }
 
 func (ps *ProvisioningService) Init() error {
-	datasourcePath := path.Join(ps.Cfg.ProvisioningPath, "datasources")
-	if err := datasources.Provision(datasourcePath); err != nil {
-		return fmt.Errorf("Datasource provisioning error: %v", err)
+	ps.log = log.New("provisioning")
+	ps.dashProvisionerChan = make(chan *dashboards.DashboardProvisioner, 1)
+	ps.pauseChan = make(chan interface{})
+
+	err := ps.ProvisionDatasources()
+	if err != nil {
+		return err
 	}
 
-	alertNotificationsPath := path.Join(ps.Cfg.ProvisioningPath, "notifiers")
-	if err := notifiers.Provision(alertNotificationsPath); err != nil {
-		return fmt.Errorf("Alert notification provisioning error: %v", err)
+	err = ps.ProvisionNotifications()
+	if err != nil {
+		return err
+	}
+
+	err = ps.ProvisionDashboards()
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
 func (ps *ProvisioningService) Run(ctx context.Context) error {
+	for {
+		select {
```

Comment:
Looks a bit crude to me but did not figure out anything better till now. We want to exit if root context got cancelled. We also want to cancel the polling at some point when reloading and allow to restart it with new config. At the same time the reload api should return when the new load is done so that part should be sync and at the same time there should not be race condition with the polling from old config.

That seems to need 2 step synchronization. When reloading we need to stop polling to prevent race condition. Then read new config do initial provisioning and finally start polling based on new config. Here `pauseChan` stops the polling to restart the polling send new provisioner in `dashProvisionerChan` and polling will start for it.